### PR TITLE
[air] Add tutorial to convert existing pytorch code to Ray AIR

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This adds a tutorial how to convert existing PyTorch training code to Ray AIR.

Please use the MyST file for review.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+To facilitate this, we only need a few changes to the code:
+
+1. We import Ray Train:
+
+```python
+import ray.train as train
+```
+
+
+2. We use a `config` dict to configure some hyperparameters (this is not strictly needed but good practice):
```
Comment (Yard1) [inline doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:242]:
```suggestion
2. We use a `config` dict to configure some hyperparameters (this is not strictly needed but good practice, especially if you intend to do hyperparameter optimization later):
```

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+    test_dataloader = train.torch.prepare_data_loader(test_dataloader)
+```
+
+5. We prepare the model for distributed gradient updates:
+
+```python
+    model = train.torch.prepare_model(model)
+```
+
+Note that `train.torch.prepare_model()` also automatically takes care of setting up devices (e.g. GPU training) - so we can get rid of those lines in our current code!
```
Comment (Yard1) [inline doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:270]: Should we explain that this wraps the model in DDP?

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
++++
+
+### Enabling checkpointing to retrieve the model
+Enabling checkpointing is pretty easy - we just need to call the `train.save_checkpoint()` API and pass the model state to it:
+
+```python
+    train.save_checkpoint(epoch=t, model=model.module.state_dict())
+```
+
+Note that the `model.module` part is needed because the model gets wrapped in `torch.nn.DistributedDataParallel`.
```
Comment (Yard1) [inline doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:401]:
```suggestion
Note that the `model.module` part is needed because the model gets wrapped in `torch.nn.DistributedDataParallel` by `train.torch.prepare_model`.
```

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+
+### Move the data loader to the training function
+
+You may have noticed a warning: `Warning: The actor TrainTrainable is very large (52 MiB). Check that its definition is not implicitly capturing a large array or other object in scope. Tip: use ray.put() to put large objects in the Ray object store.`.
+
+This is because we load the data outside the training function. Ray then serializes it to make it accessible to the remote tasks (that may get executed on a remote node!). This is not too bad with just 52 MB of data, but imagine this were a full image dataset - you wouldn't want to ship this around the cluster unnecessarily. Instead, you should move the dataset loading part into the `train_func()`. This will then download the data *to disk* once per machine and result in much more efficient data loading.
+
+The result looks like this:
```
Comment (Yard1) [inline doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:409]: This is great!

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+However, we can also use Ray AIRs `BatchPredictor` class to do scalable prediction.
+
+```{code-cell} ipython3
+from ray.air import BatchPredictor
+from ray.air.predictors.integrations.torch import TorchPredictor
+
+batch_predictor = BatchPredictor.from_checkpoint(result.checkpoint, TorchPredictor, model=NeuralNetwork())
+```
+
+The Batch predictors work with Ray Datasets. Here we convert our test dataset into a Ray Dataset - note that this is not very efficient, and you can look at our [other tutorials](https://docs.ray.io/en/master/ray-air/examples/index.html) to see more efficient ways to generate a Ray Dataset.
```
Comment (Yard1) [inline doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:555]:
```suggestion
Batch Predictors work with Ray Datasets. Here we convert our test dataset into a Ray Dataset - note that this is not very efficient, and you can look at our [other tutorials](https://docs.ray.io/en/master/ray-air/examples/index.html) to see more efficient ways to generate a Ray Dataset.
```

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+## Summary
+
+This tutorial demonstrated how to turn your existing PyTorch code into code you can use with Ray AIR.
+
+We learned how to
+- enable distributed training using Ray Train abstractions
+- save and retrieve model checkpoints via Ray AIR
+- load a model for batch prediction
+
+In our [other examples](https://docs.ray.io/en/master/ray-air/examples/index.html) you can learn how to do more things with the Ray AIR API, such as **serving your model with Ray Serve** or **tune your hyperparameters with Ray Tune.** You can also learn how to **construct Ray Datasets** to leverage Ray AIR's **preprocessing** API.
```
Comment (Yard1) [inline doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:605]: can we use a sphinx ref instead?

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md (hunk 5 above)
Comment (Yard1) [reply doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:555]: Can we also use a sphinx ref?

**Review by Yard1 — APPROVED** [review]
Looks awesome! Just some nits

File: doc/source/_toc.yml [hunk 7]
```diff
@@ -25,6 +25,7 @@ parts:
     - file: ray-air/examples/index
       sections:
         - file: ray-air/examples/torch_image_example
+        - file: ray-air/examples/convert_existing_pytorch_code_to_ray_air
```
Comment (richardliaw) [inline doc/source/_toc.yml:28]: please also add it to the index file!

File: doc/source/ray-air/examples/index.rst [hunk 8]
```diff
@@ -13,6 +13,11 @@ Framework-specific Examples
 - :doc:`/ray-air/examples/sklearn_example`: Integrating with Scikit-Learn (non-distributed)
 
 
+Starter guides
+--------------
+- :doc:`/ray-air/examples/convert_existing_pytorch_code_to_ray_air`
```
Comment (richardliaw) [inline doc/source/ray-air/examples/index.rst:18]: merge into 'framework-specific', and add a quick text description after?

File: doc/source/_toc.yml (hunk 7 above)
Comment (richardliaw) [reply doc/source/_toc.yml:28]: saw that you did, thanks

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 9]
```diff
@@ -0,0 +1,607 @@
...
+
+# Convert existing PyTorch code to Ray AIR
+
+If you already have working PyTorch code, you don't have to start from scratch to utilize the benefits of Ray Air. Instead, you can continue to use your existing code and incrementally add Ray AIR components as needed.
+
+Some of the benefits you'll get by using Ray AIR with your existing PyTorch training code:
+
+- Easy distributed data-parallel training on a cluster
+- Automatic checkpointing/fault tolerance and result tracking
+- Parallell data preprocessing
```
Comment (richardliaw) [inline doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:22]:
```suggestion
- Parallel data preprocessing
```

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+    jupytext_version: 1.13.6
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Convert existing PyTorch code to Ray AIR
+
+If you already have working PyTorch code, you don't have to start from scratch to utilize the benefits of Ray Air. Instead, you can continue to use your existing code and incrementally add Ray AIR components as needed.
```
Comment (richardliaw) [inline doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:16]:
```suggestion
If you already have working PyTorch code, you don't have to start from scratch to utilize the benefits of Ray AIR. Instead, you can continue to use your existing code and incrementally add Ray AIR components as needed.
```

Caps for AIR

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md (hunk 4 above)
Comment (amogkam) [reply doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:409]: +1, love this flow!

**Review by amogkam — APPROVED** [review]
Thanks @krfricke, this looks really good!

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 11]
```diff
@@ -0,0 +1,607 @@
+---
```
[the remark this answers is not in the document]
Comment (krfricke) [reply doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:1]: No the md is just for review

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md (hunk 2 above)
Comment (krfricke) [reply doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:270]: I think we shouldn't be too verbose here, it's not about technical detail but more for hands on.

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md [hunk 12]
```diff
@@ -0,0 +1,607 @@
...
+Some of the benefits you'll get by using Ray AIR with your existing PyTorch training code:
+
+- Easy distributed data-parallel training on a cluster
+- Automatic checkpointing/fault tolerance and result tracking
+- Parallell data preprocessing
+- Seamless integration with hyperparameter tuning
+- Scalable batch prediction
+- Scalable model serving
+
+This tutorial will show you how to start with Ray AIR from your existing PyTorch training code. We will learn how to **distribute your training** and do **scalable batch prediction**.
```
[the remark this answers is not in the document]
Comment (krfricke) [reply doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md:27]: So this is more a getting started thing for existing code - I believe for follow up processing like hyperparameter tuning and serving we should guide people to other tutorials.

File: doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md (hunk 2 above)
Comment (Yard1) [reply doc/source/ray-air/examples/convert_existing_pytorch_code_to_ray_air.md]: That's fair, but people may end up wrapping the model themselves if they don't know it will already be done IMO

**Review by richardliaw — APPROVED** [review]
